src/compiler/binder.ts

```diff
@@ -537,7 +537,18 @@ namespace ts {
                 symbol.parent = parent;
             }
 
-            if (node.flags & NodeFlags.Deprecated) {
+            const hasDeprecated = !!(node.flags & NodeFlags.Deprecated);
+            if (symbol.flags & (SymbolFlags.Signature | SymbolFlags.Function | SymbolFlags.Method)) {
+                symbol.allSignatureDeprecated = hasDeprecated && (symbol.allSignatureDeprecated ?? true);
+                if (symbol.allSignatureDeprecated) {
+                    symbol.flags |= SymbolFlags.Deprecated;
+                }
+                else {
+                    symbol.flags &= ~SymbolFlags.Deprecated;
+                }
+            }
+            else if (hasDeprecated) {

```

Comment:
merged signatures/other things will end up deprecating the whole thing, right?

```ts
/** @deprecated */
interface F { }
function F() { }
```

Reply:
Later: No, the code correctly unsets the flag, but please add a test like this (or point me to an existing one).

Reply:
I had the same fear - I don't really understand why it needs to check the symbol if the resolved signature is already there though.

Reply:
Because to avoid duplicate diagnostic. eg:
```
/* @deprecated */
declare function foo (); void;

foo;
~~~ // foo is deprecated. checked by identifier

foo();
~~~ // foo is deprecated. checked by identifier
~~~~~ // (): void is deprecated. checked by callExpression
```